### ISSUE TITLE
[ENH] Support packed C structs in pure Python mode

### ISSUE THREAD

**Issue by scoder** [cython/cython#4253]
Provide a Python syntax version of
```cython
cdef packed struct MyStruct: ...
```
Since `cython.struct()` already takes arbitrary keyword arguments, a new keyword argument `packed=True` will not work.
We could just add a new function `cython.packed_struct()`. (It would just be an alias for `struct()` in `Shadow.py`.)

**Comment by 0dminnimda** [comment]
Can we afford to make the structures a little more pythonic? Dataclass-style, only not allowing any other members besides fields.
This would look prettier:
```python
@cython.struct
class MyStruct:
    x: cython.int
    y: cython.int
    data: cython.double
```

Than this:
```python
MyStruct = cython.struct(x=cython.int, y=cython.int, data=cython.double)

```

And then it will be possible to write the packed structures in this way:
```python
@cython.packed
@cython.struct
class MyStruct:
    x: cython.int
    y: cython.int
    data: cython.double
```

or

```python
@cython.packed_struct
class MyStruct:
    x: cython.int
    y: cython.int
    data: cython.double
```

**Comment by scoder** [comment]
Or even `class MyStruct(metaclass=cython.struct)` *half-wink*

A decorator `@cython.packed` doesn't seem a good idea to me. There is nothing you could combine it with than `@cython.struct`. And `@cython.packed_struct` reads a lot more like the C code that it aims to represent.

Defining C structs in a Python class body seems a very obvious thing to do. It would be nice to support that.

**Comment by 0dminnimda** [comment]
> It would be nice to support that.

Should I create a separate issue dedicated to this?

**Comment by 0dminnimda** [comment]
> restrict use of `cython.declare` (disallow using `visibility` argument).

Of course, it's easier to just not allow visibility to be defined, but in cpp it is possible to define visibility for structure members, we could add this feature, although it seems like a relatively unclaimed feature, and it will take a lot of work to put everything together.

**Comment by scoder** [comment]
> > It would be nice to support that.
> 
> Should I create a separate issue dedicated to this?

I don't think that's necessary. If we implement it as part of this ticket, then supporting both packed and normal structs will be entirely trivial. And if we want to take the easy path first, then there's still enough time to create a new ticket.

**Comment by 0dminnimda** [comment]
Although there are some [shenanigans](https://stackoverflow.com/q/5368028/12141949) for c to recreate this behavior.

**Comment by scoder** [comment]
> > restrict use of `cython.declare` (disallow using `visibility` argument).
> 
> Of course, it's easier to just not allow visibility to be defined, but in cpp it is possible to define visibility for structure members, we could add this feature, although it seems like a relatively unclaimed feature, and it will take a lot of work to put everything together.

That seems outside of the scope of this ticket.

**Comment by 0dminnimda** [comment]
> there's still enough time to create a new ticket.

Sure


> if we want to take the easy path first

Hopefully we'll make the full version right away.

**Comment by 0dminnimda** [comment]
> That seems outside of the scope of this ticket.

I just don’t know if it’s worth raising this question at all, I don’t see a simple and backward compatible implementation ... I’m probably not thinking about it enough yet.
Anyway, should I create a separate issue this time around?

**Comment by scoder** [comment]
> should I create a separate issue this time around?

I remember discussing this topic years ago, but not where. If there isn't an issue yet, sure, would be good to have one.

**Comment by 0dminnimda** [comment]
> Or even `class MyStruct(metaclass=cython.struct)`

To also raise exceptions in pure python mode in the process of creating a class in case of an attempt to use prohibited members?

**Comment by scoder** [comment]
>> Or even `class MyStruct(metaclass=cython.struct)`
> 
> To also raise exceptions in pure python mode in the process of creating a class in case of an attempt to use prohibited members?

I was more thinking that the result of the class creation should be a
Python Enum when running in Python. A metaclass could avoid the actual
Python class creation.

Anyway, a decorator seems a much more natural and user friendly concept
than a metaclass.

**Comment by 0dminnimda** [comment]
> Python Enum when running in Python

Python Enum for `cython.struct`?

**Comment by 0dminnimda** [comment]
> A metaclass could avoid the actual
> Python class creation.
> 
> Anyway, a decorator seems a much more natural and user friendly concept
> than a metaclass.

With the help of a decorator, we can make a class that uses a metaclass, so everything you listed is still possible.
```python
def decorator(cls):
    class Wrapper(cls, metaclass=type):
        pass
    return Wrapper
```

**Comment by scoder** [comment]
> > Python Enum when running in Python
> 
> Python Enum for `cython.struct`?

Too many tickets open at the same time, I guess…

**Comment by scoder** [comment]
>> A metaclass could avoid the actual Python class creation.
>> Anyway, a decorator seems a much more natural and user friendly concept than a metaclass.
> 
> With the help of a decorator, we can make a class that uses a metaclass, so everything you listed is still possible.
> 
> ```python
> def decorator(cls):
>     class Wrapper(cls, metaclass=type):
>         pass
>     return Wrapper
> ```

No, not if the intention is to change the class creation. The decorated class is already created when the decorator runs on it.

**Comment by h-vetinari** [comment]
> using metaclasses for anything would be a totally new concept in Cython's pure Python mode

Now _that_ sounded like it would be deserving of its own issue, but it seems like there's nothing new under the sun: #1332 😅

**Comment by 0dminnimda** [comment]
> A decorator `@cython.packed` doesn't seem a good idea to me. There is nothing you could combine it with than `@cython.struct`

In the first solution from #4257, we could combine `cython.struct` with `cython.cppstruct` and `cython.struct`

**Comment by scoder** [comment]
> > A decorator `@cython.packed` doesn't seem a good idea to me. There is nothing you could combine it with than `@cython.struct`
> 
> In the first solution from #4257, we could combine `cython.packed` with `cython.cppstruct` and `cython.struct`

Still, packed structs are a very rare and special thing. I think it's fine to give them their own, separate decorator. Or two, if you want C++ structs.

I just think that
```cython
@cython.packed_struct
class MyStruct: ...
```
reads better than the more verbose
```cython
@cython.packed
@cython.struct
class MyStruct: ...
```
or
```cython
@cython.struct
@cython.packed
class MyStruct: ...
```
In which order would you write the two decorators? Is it "`@packed @struct`" or "`@struct @packed`"? Should that matter? Should people have to decide this each time they write a packed struct? If it's "`@packed @struct`", as you would find it in C, then does that make the "`@struct`" harder to spot? What if multiple developers in the same project feel differently about it?

To me, all of this seems to violate the "don't make me think" principle, big time. With "`@cython.packed_struct`", there is no need to discuss or consider anything.

**Comment by 0dminnimda** [comment]
Well, I agree. Okay, so be it.
